File: backend/app/schemas/model_config.py
"""
模型配置相关Schema
支持用户动态选择生成模型和评审模型
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from enum import Enum
from datetime import datetime


class ModelProvider(str, Enum):
    """模型提供商"""
    DEEPSEEK = "deepseek"
    MOONSHOT = "moonshot"
    QWEN = "qwen"
    OPENAI = "openai"
    OLLAMA = "ollama"  # Ollama 本地模型
    CUSTOM = "custom"  # 用户自定义模型


class ModelConfig(BaseModel):
    """模型配置"""
    provider: ModelProvider = Field(..., description="模型提供商")
    model: Optional[str] = Field(None, description="具体模型名称，不填则使用默认")
    api_key: Optional[str] = Field(None, description="API Key，不填则使用系统配置")
    base_url: Optional[str] = Field(None, description="API Base URL，不填则使用默认")
    custom_id: Optional[int] = Field(None, description="自定义模型ID，仅当provider为custom时使用")
    
    class Config:
        json_schema_extra = {
            "example": {
                "provider": "deepseek",
                "model": "deepseek-chat",
                "api_key": None,
                "base_url": None
            }
        }


class ModelConfigRequest(BaseModel):
    """模型配置请求"""
    # 需求分析模型配置（用于需求分析阶段）
    requirement_analyze_model: Optional[ModelConfig] = Field(
        None,
        description="需求分析模型配置"
    )

    # 用例生成模型配置（用于用例生成）
    testcase_generate_model: Optional[ModelConfig] = Field(
        None,
        description="用例生成模型配置"
    )

    # 用例评审模型配置（用于用例评审、定稿）
    testcase_review_model: Optional[ModelConfig] = Field(
        None,
        description="用例评审模型配置"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "requirement_analyze_model": {
                    "provider": "deepseek",
                    "model": "deepseek-chat"
                },
                "testcase_generate_model": {
                    "provider": "deepseek",
                    "model": "deepseek-chat"
                },
                "testcase_review_model": {
                    "provider": "moonshot",
                    "model": "moonshot-v1-32k"
                }
            }
        }


class AvailableModel(BaseModel):
    """可用模型信息"""
    provider: ModelProvider
    model_name: str
    display_name: str
    description: str
    supported_features: List[str] = Field(default_factory=list)
    is_available: bool = Field(default=True, description="是否已配置API Key")
    custom_id: Optional[int] = Field(None, description="自定义模型ID")


class ModelListResponse(BaseModel):
    """可用模型列表响应"""
    requirement_analyze_models: List[AvailableModel] = Field(default_factory=list)
    testcase_generate_models: List[AvailableModel] = Field(default_factory=list)
    testcase_review_models: List[AvailableModel] = Field(default_factory=list)
    default_requirement_analyze: ModelProvider = ModelProvider.DEEPSEEK
    default_testcase_generate: ModelProvider = ModelProvider.DEEPSEEK
    default_testcase_review: ModelProvider = ModelProvider.MOONSHOT


# ========== 自定义模型相关Schema ==========

class CustomModelCreate(BaseModel):
    """创建自定义模型请求"""
    name: str = Field(..., description="模型标识（如 kimi）", min_length=1, max_length=50)
    display_name: str = Field(..., description="显示名称（如 Kimi AI）", min_length=1, max_length=100)
    base_url: str = Field(..., description="API 地址（如 https://api.moonshot.cn/v1）", min_length=1, max_length=500)
    api_key: str = Field(..., description="API Key", min_length=1, max_length=500)
    default_model: str = Field(..., description="默认模型名", min_length=1, max_length=100)
    description: Optional[str] = Field(None, description="模型描述", max_length=500)


class CustomModelUpdate(BaseModel):
    """更新自定义模型请求"""
    display_name: Optional[str] = Field(None, description="显示名称", max_length=100)
    base_url: Optional[str] = Field(None, description="API 地址", max_length=500)
    api_key: Optional[str] = Field(None, description="API Key", max_length=500)
    default_model: Optional[str] = Field(None, description="默认模型名", max_length=100)
    description: Optional[str] = Field(None, description="模型描述", max_length=500)
    enabled: Optional[bool] = Field(None, description="是否启用")


class CustomModelResponse(BaseModel):
    """自定义模型响应"""
    id: int
    name: str
    display_name: str
    base_url: str
    default_model: str
    description: Optional[str] = None
    enabled: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/api/modelConfig.ts
/**
 * 模型配置 API
 */
import api from './index'

// ============ 旧版接口（保留向后兼容）============

export interface AvailableModel {
  provider: string
  model_name: string
  display_name: string
  description: string
  supported_features: string[]
  is_available: boolean
}

export interface ModelListResponse {
  requirement_analyze_models: AvailableModel[]
  testcase_generate_models: AvailableModel[]
  testcase_review_models: AvailableModel[]
  default_requirement_analyze: string
  default_testcase_generate: string
  default_testcase_review: string
}

export interface ModelProvider {
  name: string
  display_name: string
  description: string
  configured: boolean
  default_model: string
}

// ============ 新版接口（数据库配置）============

export interface LLMModel {
  id: number
  name: string
  display_name: string
  provider: string
  base_url: string
  api_key: string | null
  default_model: string
  description: string | null
  enabled: boolean
  is_default: boolean
  created_at: string | null
  updated_at: string | null
}

export interface EmbeddingModel {
  id: number
  name: string
  display_name: string
  provider: string
  api_base: string
  api_key: string | null
  model_name: string
  dimension: number
  description: string | null
  enabled: boolean
  is_default: boolean
  created_at: string | null
  updated_at: string | null
}

export interface ModelStatus {
  requirement_analyze: {
    configured: boolean
    model: string | null
    model_name: string | null
    provider: string | null
    id: number | null
    is_custom: boolean
  }
  testcase_generate: {
    configured: boolean
    model: string | null
    model_name: string | null
    provider: string | null
    id: number | null
    is_custom: boolean
  }
  testcase_review: {
    configured: boolean
    model: string | null
    model_name: string | null
    provider: string | null
    id: number | null
    is_custom: boolean
  }
  embedding: {
    configured: boolean
    model: string | null
    model_name: string | null
    provider: string | null
    dimension: number | null
  }
  all_required_configured: boolean
  rag_available: boolean
}

export interface ModelStatusSummary {
  configured_count: number
  total_count: number
  all_configured: boolean
  missing_models: Array<{
    purpose: string
    display_name: string
  }>
}

export const modelConfigApi = {
  // ============ 旧版接口（保留向后兼容）============

  /**
   * 获取可用模型列表
   */
  getAvailableModels: () => {
    return api.get<ModelListResponse>('/models/available')
  },

  /**
   * 获取模型提供商列表
   */
  getModelProviders: () => {
    return api.get<{ providers: ModelProvider[] }>('/models/providers')
  },

  // ============ 新版接口（数据库配置）============

  // LLM 模型配置
  getLLMModels: () => {
    return api.get<LLMModel[]>('/v1/llm-models')
  },

  getEnabledLLMModels: () => {
    return api.get<LLMModel[]>('/v1/llm-models/enabled')
  },

  getLLMModel: (id: number) => {
    return api.get<LLMModel>(`/v1/llm-models/${id}`)
  },

  createLLMModel: (data: Partial<LLMModel>) => {
    return api.post<{ id: number; name: string; message: string }>('/v1/llm-models', data)
  },

  updateLLMModel: (id: number, data: Partial<LLMModel>) => {
    return api.put<{ id: number; name: string; message: string }>(`/v1/llm-models/${id}`, data)
  },

  deleteLLMModel: (id: number) => {
    return api.delete<{ message: string }>(`/v1/llm-models/${id}`)
  },

  testLLMConnection: (data: { base_url: string; api_key?: string; model: string; provider?: string }) => {
    return api.post<{ success: boolean; message: string }>('/v1/llm-models/test', data)
  },

  // 测试指定模型连接（后端内部获取真实api_key）
  testLLMModelById: (modelId: number) => {
    return api.post<{ success: boolean; message: string }>(`/v1/llm-models/${modelId}/test`)
  },

  // 获取模型详情（包含真实api_key，用于测试）
  getLLMModelForTest: (id: number) => {
    return api.get<{ id: number; name: string; display_name: string; provider: string; base_url: string; api_key?: string; default_model: string }>(`/v1/llm-models/${id}/for-test`)
  },

  getLLMModelDefaults: () => {
    return api.get<Array<{ purpose: string; model_name: string; model_type: string; updated_at: string }>>('/v1/llm-models/defaults/all')
  },

  setDefaultModel: (purpose: string, model_name: string) => {
    return api.put<{ message: string; purpose: string; model_name: string }>(`/v1/llm-models/defaults/${purpose}`, { model_name })
  },

  getModelForPurpose: (purpose: string) => {
    return api.get<{ configured: boolean; model?: any; message?: string }>(`/v1/llm-models/purpose/${purpose}`)
  },

  // 嵌入模型配置
  getEmbeddingModels: () => {
    return api.get<EmbeddingModel[]>('/v1/embedding-models')
  },

  getEnabledEmbeddingModels: () => {
    return api.get<EmbeddingModel[]>('/v1/embedding-models/enabled')
  },

  getDefaultEmbeddingModel: () => {
    return api.get<{ configured: boolean; model?: any; message?: string }>('/v1/embedding-models/default')
  },

  getEmbeddingModel: (id: number) => {
    return api.get<EmbeddingModel>(`/v1/embedding-models/${id}`)
  },

  createEmbeddingModel: (data: Partial<EmbeddingModel>) => {
    return api.post<{ id: number; name: string; message: string }>('/v1/embedding-models', data)
  },

  updateEmbeddingModel: (id: number, data: Partial<EmbeddingModel>) => {
    return api.put<{ id: number; name: string; message: string }>(`/v1/embedding-models/${id}`, data)
  },

  deleteEmbeddingModel: (id: number) => {
    return api.delete<{ message: string }>(`/v1/embedding-models/${id}`)
  },

  testEmbeddingConnection: (data: { api_base: string; api_key: string; model_name: string; provider?: string; dimension?: number }) => {
    return api.post<{ success: boolean; message: string }>('/v1/embedding-models/test', data)
  },

  setDefaultEmbeddingModel: (name: string) => {
    return api.put<{ message: string; name: string }>(`/v1/embedding-models/default/${name}`)
  },

  // 模型状态
  getModelStatus: () => {
    return api.get<ModelStatus>('/v1/model-status')
  },

  getModelStatusSummary: () => {
    return api.get<ModelStatusSummary>('/v1/model-status/summary')
  },

  initializeModels: () => {
    return api.post<{ success: boolean; message: string; llm_models_count?: number; embedding_configs_count?: number }>('/v1/model-status/initialize')
  },
}
